content/learning-paths/mobile-graphics-and-gaming/android_halide/fusion.md

```diff
@@ -0,0 +1,349 @@
+Halide::Buffer<uint8_t> blurBuffer = blur.realize({ width, height });
+
+Halide::Func thresholded("thresholded");
+thresholded(x, y) = Halide::cast<uint8_t>(Halide::select(blurBuffer(x, y) > 128, 255, 0));
+
+Halide::Buffer<uint8_t> outputBuffer = thresholded.realize({ width, height });
+```
+
+### Performance profiling to identify bottlenecks
+The primary bottleneck in the non-fused pipeline is memory access:
+1. Intermediate results (blurBuffer) are written to and read from memory.
+2. This results in additional latency, reduced cache efficiency, and unnecessary memory bandwidth usage.
+
+Profiling this pipeline with tools like Halide’s built-in profiler typically shows memory bandwidth as a major limiting factor. 
+For details on how to enable and interpret Halide’s profiler, please refer to the official [Halide profiling tutorial](https://halide-lang.org/tutorials/tutorial_lesson_21_auto_scheduler_generate.html#profiling).
+
+## Applying operation fusion in Halide
+To apply operation fusion, Halide provides powerful scheduling constructs that allow you to precisely control when and where operations are computed and stored.
+
+### Scheduling techniques
+The three primary Halide scheduling methods to enable fusion are:
+1. compute_at - compute the values of one Func at the iteration point of another.
+2. store_at - store intermediate results at a particular loop iteration or stage to minimize memory footprint.
+3. fuse - merge loop variables of two dimensions into one, improving loop efficiency.
+
+### Using constructs like compute_at, store_at, and fuse
+Let’s fuse the blur and thresholding stages using compute_at to ensure both operations execute together, eliminating intermediate storage. To do so, create a new file camera-capture-fusion.cpp, and modify it as follows:
+
+```cpp
+#include "Halide.h"
+#include <opencv2/opencv.hpp>
+#include <iostream>
+#include <exception>
+
+using namespace cv;
+using namespace std;
+
+// This function clamps the coordinate (coord) within [0, maxCoord - 1].
+static inline Halide::Expr clampCoord(Halide::Expr coord, int maxCoord) {
+    return Halide::clamp(coord, 0, maxCoord - 1);
+}
+
+int main() {
+    // Open the default camera with OpenCV.
+    VideoCapture cap(0);
+    if (!cap.isOpened()) {
+        cerr << "Error: Unable to open camera." << endl;
+        return -1;
+    }
+
+    while (true) {
+        // Capture a frame from the camera.
+        Mat frame;
+        cap >> frame;
+        if (frame.empty()) {
+            cerr << "Error: Received empty frame." << endl;
+            break;
+        }
+
+        // Convert the frame to grayscale.
+        Mat gray;
+        cvtColor(frame, gray, COLOR_BGR2GRAY);
+
+        // Ensure the grayscale image is continuous in memory.
+        if (!gray.isContinuous()) {
+            gray = gray.clone();
+        }
+
+        int width  = gray.cols;
+        int height = gray.rows;
+
+        // Create a simple 2D Halide buffer from the grayscale Mat.
```

Comment:
I think the Halide code below could be modified in the most recommended way which is mentioned in the previous section